### PR TITLE
Fix exports and clean up API

### LINKED ISSUE #58: Clean up API

### PR DESCRIPTION
**Pull request by lhorie**
Resolves #58 

Motivation:

There were numerous issues with the plugin:

- Tokens were not exported
- Token names were inconsistent with other packages
- Docs were incomplete
- It is weird to require every consumer of this package to create their own token in order to be able to use the package at all. The extra user-space token makes it harder to codemod and adds another possible point of failure for future automated migrations.

  The aliasing mechanism is also weird and confusing to understand, and overlaps in functionality with `app.enhance`.

```js
//before
import {FetchToken, createToken} from 'fusion-tokens';
import CsrfProtection from 'fusion-plugin-csrf-protection';

const BaseFetchToken = createToken('BaseFetchToken');
app.register(BaseFetchToken, fetch);
app.register(FetchToken, CsrfProtection).alias(FetchToken, BaseFetchToken);
```

```js
//after
import CsrfProtection, {FetchForCsrfToken} from 'fusion-plugin-csrf-protection';

app.register(FetchForCsrfToken, fetch);
app.register(FetchToken, CsrfProtection);
```

Resolution:

This PR exports all tokens, and renamed them to be consistent with other plugins. It also made the need for an extra user-space token obsolete

### CODE REVIEW

**Comment by ganemone** [comment]
Aliasing was built before `app.enhance`. `app.enhance` is designed to replace aliasing

File: src/index.js [hunk 1]
```diff
@@ -8,3 +8,5 @@ import serverCsrf from './server';
 import clientCsrf from './browser';
 
 export default (__NODE__ ? serverCsrf : clientCsrf);
+
+export {FetchToken} from './tokens';
```
Comment (dennisgl) [inline src/index.js]: typo? `FetchForCsrfToken`

**Comment by AlexMSmithCA** [comment]
I'm worried that this sets a bad precedent moving forward, where plugin authors can export their own "special-for-me" tokens for dependencies that should be more generic.  Isn't the _enhance_ functionality built around this use case (a la enhancing _fetch_)?

If you _really_ want to override the polymorphism aspect of IoC, we could add the ability to inject specific dependencies for tokens as opposed to aliasing together two tokens:

`app.register(FetchToken, CsrfProtection).overrideDependency(FetchToken, fetch)`

**Comment by lhorie** [comment]
We already have a lot special-for-me tokens (e.g. all the config ones)

WRT `.overrideDependency`, we had discussed a limitation of that approach. The problem is that you sometimes need to override something with a value that only exists within the DI container (e.g. in the complex version of this case, you'll want CsrfProtection to use fetch, but then you want RoutePrefix to use CsrfProtection's fetch.

**Comment by AlexMSmithCA** [comment]
The primary distinction is whether we believe the app author or the plugin author should be the one to designate the dependency "override".  If it is the plugin author to the extent that the dependency being provided should be different from any other dependency, then I think that makes sense.

It is likely that we will still require an override method for the case when the userland author wishes to override a dependency (as opposed to the plugin author).

As an aside, the `.overrideDependency` case would not need any additional tokens (ad-hoc or provided by plugins).  I think the primary drawback, however, is that does introduce a new API endpoint, and one which supersedes the IoC graph - and so may be better as a replacement proposal for `.alias`.

**Comment by lhorie** [comment]
I think it's up to the plugin author to decide what dependencies the plugin should receive. There isn't a particularly good reason for CsrfProtection to depend on `FetchToken` specifically, if the intended consumption is to register `CsrfProtection` _to_ `FetchToken`.

In user space, they already have the ability to "override" things: using the normal registration mechanism. I actually think `override` and `enhance` are the wrong abstractions. The use case for `alias` is "A and B both depend on LoggerToken, but we want to pass different loggers to each". Neither `enhance` nor `override` serve that use case. The use case of "A and B depend on FetchToken and we want FetchToken to mean some composition of fetch" can easily be achieved the way this PR proposes, with no alias, enhance or override shenanigans.

**Comment by AlexMSmithCA** [comment]
+1, that sounds reasonable to me.  We could be cognizant of which tokens we publish as plugin authors to avoid the case where each plugin has only plugin-specific tokens (e.g. "LoggerForPluginA") when the general case would be sufficient.  In this case, the distinction is appropriate given that there is a clear distinction between `FetchToken` and `FetchForCsrfToken` and they should *not* be the same implementation.

**Comment by lhorie** [comment]
!merge

**Pull request merged**